crates/ruff_linter/src/rules/ruff/rules/unnecessary_nested_literal.rs

```diff
@@ -0,0 +1,129 @@
+use ruff_diagnostics::{Applicability, Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{Expr, ExprContext, ExprSubscript, ExprTuple};
+use ruff_python_semantic::analyze::typing::traverse_literal;
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for unnecessary nested `Literal`.
+///
+/// ## Why is this bad?
+/// Parameterization of literals by other literals is supported as an ergonomic
+/// feature as proposed in [PEP 586], to enable patterns such as:
+/// ```python
+/// ReadOnlyMode         = Literal["r", "r+"]
+/// WriteAndTruncateMode = Literal["w", "w+", "wt", "w+t"]
+/// WriteNoTruncateMode  = Literal["r+", "r+t"]
+/// AppendMode           = Literal["a", "a+", "at", "a+t"]
+///
+/// AllModes = Literal[ReadOnlyMode, WriteAndTruncateMode,
+///                   WriteNoTruncateMode, AppendMode]
+/// ```
+///
+/// As a consequence, type checkers also support nesting of literals
+/// which is less readable than a flat `Literal`:
+/// ```python
+/// AllModes = Literal[Literal["r", "r+"], Literal["w", "w+", "wt", "w+t"],
+///                   Literal["r+", "r+t"], Literal["a", "a+", "at", "a+t"]]
+/// ```
+///
+/// ## Example
+/// ```python
+/// AllModes = Literal[
+///     Literal["r", "r+"],
+///     Literal["w", "w+", "wt", "w+t"],
+///     Literal["r+", "r+t"],
+///     Literal["a", "a+", "at", "a+t"],
+/// ]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// AllModes = Literal[
+///     "r", "r+", "w", "w+", "wt", "w+t", "r+", "r+t", "a", "a+", "at", "a+t"
+/// ]
+/// ```
+///
+/// or assign the literal to a variable as in the first example.
+///
+/// ## Fix safety
+/// The fix for this rule is marked as unsafe when the `Literal` contains comments.
+///
+/// ## References
+/// - [Typing documentation: Legal parameters for `Literal` at type check time](https://typing.readthedocs.io/en/latest/spec/literal.html#legal-parameters-for-literal-at-type-check-time)
+///
+/// [PEP 586](https://peps.python.org/pep-0586/)
+#[violation]
+pub struct UnnecessaryNestedLiteral;
+
+impl Violation for UnnecessaryNestedLiteral {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Unnecessary nested `Literal`".to_string()
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Replace with flattened `Literal`".to_string())
+    }
+}
+
+/// RUF039
+pub(crate) fn unnecessary_nested_literal<'a>(checker: &mut Checker, literal_expr: &'a Expr) {
+    let mut nodes: Vec<&Expr> = Vec::new();
+    let mut is_nested = false;
+
+    let mut check_for_duplicate_members = |expr: &'a Expr, parent: &'a Expr| {
+        // If the parent is not equal to the `literal_expr` then we know we are traversing recursively.
+        if parent != literal_expr {
```

Comment:
You could consider using `AnyNodeRef::ptr_eq` here because all you want to test is if the expressions are the same (`Expr::eq` does a deep comparison by default)